File: typescript/lib/services/service_context.ts
import { ApiContainerServiceClient } from '../../kurtosis_core_rpc_api_bindings/api_container_service_grpc_pb'; 
import { ExecCommandArgs, ExecCommandResponse, FileGenerationOptions, GenerateFilesArgs, GenerateFilesResponse, LoadStaticFilesArgs, LoadStaticFilesResponse } from '../../kurtosis_core_rpc_api_bindings/api_container_service_pb';
import { ServiceID } from './service';
import { StaticFileID } from './container_creation_config'; 
import { newGetExecCommandArgs, newGetGenerateFilesArgs, newGetFileGenerationOptions, newGetLoadStaticFilesArgs } from "../constructor_calls";
import { okAsync, errAsync, ResultAsync, ok, err, Result } from 'neverthrow';
import * as grpc from "grpc";
import * as path from "path";


// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
class GeneratedFileFilepaths {
    
    private readonly absoluteFilepathHere: string;
    private readonly absoluteFilepathOnServiceContainer: string;

    constructor (
        absoluteFilepathHere: string,
        absoluteFilepathOnServiceContainer: string) {
            this.absoluteFilepathHere = absoluteFilepathHere;
            this.absoluteFilepathOnServiceContainer = absoluteFilepathOnServiceContainer;
    }   

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public getAbsoluteFilepathHere(): string { 
        return this.absoluteFilepathHere;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public getAbsoluteFilepathOnServiceContainer(): string { 
        return this.absoluteFilepathOnServiceContainer;
    }
}



class ServiceContext {
    
    private readonly client: ApiContainerServiceClient;
    private readonly serviceId: ServiceID;
    private readonly ipAddress: string;
    private readonly enclaveDataVolMountpointHere: string;
    private readonly enclaveDataVolMountpointOnServiceContainer: string;

    constructor(
        client: ApiContainerServiceClient,
        serviceId: ServiceID,
        ipAddress: string,
        enclaveDataVolMountpointHere: string,
        enclaveDataVolMountpointOnServiceContainer: string) {
            this.client = client;
            this.serviceId = serviceId;
            this.ipAddress = ipAddress;
            this.enclaveDataVolMountpointHere = enclaveDataVolMountpointHere;
            this.enclaveDataVolMountpointOnServiceContainer = enclaveDataVolMountpointOnServiceContainer;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public getServiceID(): ServiceID { 
        return this.serviceId;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public getIPAddress(): string {
        return this.ipAddress;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public async execCommand(command: string[]): Promise<Result<[number, Uint8Array | string], Error>> {
        const serviceId: ServiceID = this.serviceId;
        const args: ExecCommandArgs = newGetExecCommandArgs(serviceId, command);

        const promiseExecCommand: Promise<ResultAsync<ExecCommandResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.execCommand(args, (error: grpc.ServiceError, response: ExecCommandResponse) => {
                if (error) {
                    resolve(errAsync(error));
                } else {
                    resolve(okAsync(response));
                }
            })
        });

        const resultExecCommand: Result<ExecCommandResponse, Error> = await promiseExecCommand;

        if (!resultExecCommand.isOk()) {
            return err(resultExecCommand.error);
        }
        const resp: ExecCommandResponse = resultExecCommand.value;
        return ok([resp.getExitCode(), resp.getLogOutput()]);
        

    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public async generateFiles(filesToGenerateSet: Set<string>): Promise<Result<Map<string, GeneratedFileFilepaths>, Error>> {
        const serviceId: ServiceID = this.serviceId;
        const fileGenerationOpts: Map<string, FileGenerationOptions> = newGetFileGenerationOptions(filesToGenerateSet);

        const args: GenerateFilesArgs = newGetGenerateFilesArgs(serviceId, fileGenerationOpts);

        const promiseGenerateFiles: Promise<ResultAsync<GenerateFilesResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.generateFiles(args, (error: grpc.ServiceError, response: GenerateFilesResponse) => {
                if (error) {
                    resolve(errAsync(error));
                } else {
                    resolve(okAsync(response));
                }
            })
        });

        const resultGenerateFiles: Result<GenerateFilesResponse, Error> = await promiseGenerateFiles;

        if (!resultGenerateFiles.isOk()) {
            return err(resultGenerateFiles.error);
        } 
        const resp: GenerateFilesResponse = resultGenerateFiles.value;

        const generatedFileRelativeFilepaths: Map<string, string> = resp.getGeneratedFileRelativeFilepathsMap();

        const result: Map<string, GeneratedFileFilepaths> = new Map();
        for (let fileId in filesToGenerateSet) {
            
            var relativeFilepath: string;
            if (!generatedFileRelativeFilepaths.has(fileId)) {
                return err(new Error(
                    "No filepath (relative to test volume root) was returned for file " + fileId +  ", even though we requested it; this is a Kurtosis bug"
                    )
                );
            }
            relativeFilepath = generatedFileRelativeFilepaths[fileId];

            const absFilepathHere: string = path.join(this.enclaveDataVolMountpointHere, relativeFilepath);
            const absFilepathOnService: string = path.join(this.enclaveDataVolMountpointOnServiceContainer, relativeFilepath);
            const genFilePath: GeneratedFileFilepaths = new GeneratedFileFilepaths(absFilepathHere, absFilepathOnService);
            result.set(fileId, genFilePath); 
        }
        return ok(result);
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    public async loadStaticFiles(usedStaticFilesSet: Set<StaticFileID>): Promise<Result<Map<StaticFileID, string>, Error>> { 
        const serviceId: ServiceID = this.serviceId;
        const staticFilesToCopyStringSet: Map<string, boolean> = new Map(); 
        for (let staticFileId in usedStaticFilesSet) {
            staticFilesToCopyStringSet[String(staticFileId)] = true;
        }

        const loadStaticFilesArgs: LoadStaticFilesArgs = newGetLoadStaticFilesArgs(serviceId, staticFilesToCopyStringSet);
        
        const promiseLoadStaticFiles: Promise<ResultAsync<LoadStaticFilesResponse, Error>> = new Promise((resolve, _unusedReject) => {
            this.client.loadStaticFiles(loadStaticFilesArgs, (error: grpc.ServiceError, response: LoadStaticFilesResponse) => {
                if (error) {
                    resolve(errAsync(error));
                } else {
                    resolve(okAsync(response));
                }
            })
        });

        const resultLoadStaticFiles: Result<LoadStaticFilesResponse, Error> = await promiseLoadStaticFiles;

        if (!resultLoadStaticFiles.isOk()) {
            return err(resultLoadStaticFiles.error);
        }
        const loadStaticFilesResp: LoadStaticFilesResponse = resultLoadStaticFiles.value;

        const staticFileAbsFilepathsOnService: Map<StaticFileID, string> = new Map();
        for (let staticFileId in loadStaticFilesResp.getCopiedStaticFileRelativeFilepathsMap()) {
            const filepathRelativeToExVolRoot: string = loadStaticFilesResp.getCopiedStaticFileRelativeFilepathsMap()[staticFileId];
            const absFilepathOnContainer: string = path.join(this.enclaveDataVolMountpointOnServiceContainer, filepathRelativeToExVolRoot)
            staticFileAbsFilepathsOnService[<StaticFileID>(staticFileId)] = absFilepathOnContainer;
        }
        return ok(staticFileAbsFilepathsOnService);

    }
}

File: golang/lib/services/service_context.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package services

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-client/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-client/golang/lib/binding_constructors"
	"github.com/palantir/stacktrace"
	"path"
)

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
type GeneratedFileFilepaths struct {
	absoluteFilepathHere               string
	absoluteFilepathOnServiceContainer string
}

func newGeneratedFileFilepaths(absoluteFilepathHere string, absoluteFilepathOnServiceContainer string) *GeneratedFileFilepaths {
	return &GeneratedFileFilepaths{absoluteFilepathHere: absoluteFilepathHere, absoluteFilepathOnServiceContainer: absoluteFilepathOnServiceContainer}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (generated *GeneratedFileFilepaths) GetAbsoluteFilepathHere() string {
	return generated.absoluteFilepathHere
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (generated *GeneratedFileFilepaths) GetAbsoluteFilepathOnServiceContainer() string {
	return generated.absoluteFilepathOnServiceContainer
}



// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
type ServiceContext struct {
	client                                     kurtosis_core_rpc_api_bindings.ApiContainerServiceClient
	serviceId                                  ServiceID
	ipAddress                                  string
	enclaveDataVolMountpointHere               string
	enclaveDataVolMountpointOnServiceContainer string
}

func NewServiceContext(
		client kurtosis_core_rpc_api_bindings.ApiContainerServiceClient,
		serviceId ServiceID,
		ipAddress string,
		enclaveDataVolumeMountpointHere string,
		enclaveDataVolumeMountpointOnServiceContainer string) *ServiceContext {
	return &ServiceContext{
		client:                       client,
		serviceId:                    serviceId,
		ipAddress:                    ipAddress,
		enclaveDataVolMountpointHere: enclaveDataVolumeMountpointHere,
		enclaveDataVolMountpointOnServiceContainer:   enclaveDataVolumeMountpointOnServiceContainer,
	}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GetServiceID() ServiceID {
	return self.serviceId
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GetIPAddress() string {
	return self.ipAddress
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) ExecCommand(command []string) (int32, string, error) {
	serviceId := self.serviceId
	args := binding_constructors.NewExecCommandArgs(string(serviceId), command)
	resp, err := self.client.ExecCommand(context.Background(), args)
	if err != nil {
		return 0, "", stacktrace.Propagate(
			err,
			"An error occurred executing command '%v' on service '%v'",
			command,
			serviceId)
	}
	return resp.ExitCode, resp.LogOutput, nil
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GenerateFiles(filesToGenerateSet map[string]bool) (map[string]*GeneratedFileFilepaths, error) {
	serviceId := self.serviceId
	fileGenerationOpts := map[string]*kurtosis_core_rpc_api_bindings.FileGenerationOptions{}
	for fileId := range filesToGenerateSet {
		fileGenerationOpts[fileId] = binding_constructors.NewFileGenerationOptions(kurtosis_core_rpc_api_bindings.FileGenerationOptions_FILE)
	}
	args := binding_constructors.NewGenerateFilesArgs(string(serviceId), fileGenerationOpts)
	resp, err := self.client.GenerateFiles(context.Background(), args)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred generating files using args: %+v", args)
	}
	generatedFileRelativeFilepaths := resp.GeneratedFileRelativeFilepaths

	result := map[string]*GeneratedFileFilepaths{}
	for fileId := range filesToGenerateSet {
		relativeFilepath, found := generatedFileRelativeFilepaths[fileId]
		if !found {
			return nil, stacktrace.NewError(
				"No filepath (relative to enclave data volume root) was returned for file '%v', even though we requested it; this is a Kurtosis bug",
				fileId)
		}
		absFilepathHere := path.Join(self.enclaveDataVolMountpointHere, relativeFilepath)
		absFilepathOnService := path.Join(self.enclaveDataVolMountpointOnServiceContainer, relativeFilepath)
		result[fileId] = newGeneratedFileFilepaths(absFilepathHere, absFilepathOnService)
	}
	return result, nil
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) LoadStaticFiles(usedStaticFilesSet map[StaticFileID]bool) (map[StaticFileID]string, error) {
	serviceId := self.serviceId
	staticFilesToCopyStringSet := map[string]bool{}
	for staticFileId := range usedStaticFilesSet {
		staticFilesToCopyStringSet[string(staticFileId)] = true
	}
	loadStaticFilesArgs := binding_constructors.NewLoadStaticFilesArgs(string(serviceId), staticFilesToCopyStringSet)
	loadStaticFilesResp, err := self.client.LoadStaticFiles(context.Background(), loadStaticFilesArgs)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred loading the requested static files into the namespace of service '%v'", serviceId)
	}
	staticFileAbsFilepathsOnService := map[StaticFileID]string{}
	for staticFileId, filepathRelativeToExVolRoot := range loadStaticFilesResp.CopiedStaticFileRelativeFilepaths {
		absFilepathOnContainer := path.Join(self.enclaveDataVolMountpointOnServiceContainer, filepathRelativeToExVolRoot)
		staticFileAbsFilepathsOnService[StaticFileID(staticFileId)] = absFilepathOnContainer
	}
	return staticFileAbsFilepathsOnService, nil

}
